plugin/pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -650,46 +674,45 @@ func NewServiceAffinityPredicate(podLister algorithm.PodLister, serviceLister al
 // 	the match.
 // Otherwise:
 // 	Create an "implicit selector" which gaurantees pods will land on nodes with similar values
-//	for the affinity labels.
+// 	for the affinity labels.
+//
 // 	To do this, we "reverse engineer" a selector by introspecting existing pods running under the same service+namespace.
 //	These backfilled labels in the selector "L" are defined like so:
 // 		- L is a label that the ServiceAffinity object needs as a matching constraints.
 // 		- L is not defined in the pod itself already.
 // 		- and SOME pod, from a service, in the same namespace, ALREADY scheduled onto a node, has a matching value.
-func (s *ServiceAffinity) CheckServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+//
+// WARNING: This Predicate is NOT gauranteed to work if some of the predicateMetadata data isn't precomputed...
+// For that reason it is not exported, i.e. it is highlhy coupled to the implementation of the FitPredicate construction.
+func (s *ServiceAffinity) checkServiceAffinity(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+	var services []*api.Service
+	var pods []*api.Pod
+	if pm, ok := meta.(*predicateMetadata); ok && pm.serviceAffinityMatchingPodList != nil && pm.serviceAffinityMatchingPodServices != nil {
+		services = pm.serviceAffinityMatchingPodServices
+		pods = pm.serviceAffinityMatchingPodList
+	} else {
+		// TODO decide precomputation-data semantics, issue #33763.
+		return false, nil, fmt.Errorf("Precomputed data not available.")
```

Comment:
Well, I think that what you should do here is:
```
tmp := predicateMetadata{pod: pod}
s.serviceAffinityPrecomputation(..)
serivces, pods = tmp.serviceAff..., tmp.serviceAff...
```

With this approach, the predicate work no matter what metadata you pass - it.s just more efficient when you pass metadata. And I don't see any problems with this approach.